### PR TITLE
Validate fds created by the user

### PR DESCRIPTION
**Pull request by vinipsmaker**
write_mapping() will do the following:

openat(proc_dir_fd, map_file, O_WRONLY);

An attacker could create a directory containing a symlink named "uid_map" pointing to any file owned by root, and thus allow him to overwrite any root-owned file.

### CODE REVIEW

File: src/newgidmap.c [hunk 1]
```diff
@@ -196,6 +196,12 @@ int main(int argc, char **argv)
 		return EXIT_FAILURE;
 	}
 
+	/* Validate fds coming from get_pidfd_from_fd() */
```
Comment (hallyn) [inline src/newgidmap.c:199]: Would it, however, be better to do this in the get_pidfd_from_fd() fn itself?  I realize you have the struct stat here already, but any other callers of get_pidfd_from_fd() will need the same guarantee.

File: src/newgidmap.c [hunk 2]
```diff
@@ -196,6 +196,12 @@ int main(int argc, char **argv)
...
+	if (st.st_dev != 22 || st.st_rdev != 0) {
```
Comment (hallyn) [inline src/newgidmap.c:200]: Hm, on my laptop it's 23.  That bodes ill.

File: src/newgidmap.c (hunk 1 above)
Comment (vinipsmaker) [reply src/newgidmap.c:199]: I agree. Should I do the change?

Comment (hallyn) [reply src/newgidmap.c:199]: Yes, thanks.

File: src/newgidmap.c (hunk 2 above)
Comment (vinipsmaker) [reply src/newgidmap.c:200]: There's another way to get these values. I'll update the patchset. One sec.

Comment (vinipsmaker) [reply src/newgidmap.c]: I pushed the new version. Please, review.

Comment (hallyn) [reply src/newgidmap.c]: Looks great, thanks.

I'll take another look in 2m and then merge.

I can't tell you how embarrassed I am.

**Review by hallyn — APPROVED** [review]

**Comment by hallyn** [comment]
Thanks!

File: lib/get_pid.c [hunk 3]
```diff
@@ -51,6 +53,21 @@ int get_pidfd_from_fd(const char *pidfdstr)
...
 	}
 
+	if (stat("/proc/self/uid_map", &st) < 0) {
+		return -1;
+	}
+
+	proc_st_dev = st.st_dev;
+	proc_st_rdev = st.st_rdev;
+
+	if (fstat(proc_dir_fd, &st) < 0) {
```
Comment (vinipsmaker) [inline lib/get_pid.c:63]: there's a copy'n'paste error here. It should be `fstat(val, &st)`

Comment (hallyn) [reply lib/get_pid.c:63]: Yup, updated, thanks.

**Pull request merged**